File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/dto/CharacterDto.java
package ai.ragu.charactermanager.dto;

import ai.ragu.charactermanager.enumeration.*;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.*;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.*;


@Data
@AllArgsConstructor
@NoArgsConstructor
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@ToString(onlyExplicitlyIncluded = true)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class CharacterDto {

    @JsonProperty("id")
    @JsonPropertyDescription("A unique identifier for this character")
    @ToString.Include
    @EqualsAndHashCode.Include
    @Nonnull
    private UUID id;

    @JsonProperty("name")
    @JsonPropertyDescription("The name of this character")
    @Nonnull
    private String name;

    @JsonProperty("level")
    @JsonPropertyDescription("The level of this character")
    @Nonnull
    private Integer level = 1;

    @JsonProperty("ancestry")
    @JsonPropertyDescription("The ancestry of this character")
    @Nullable
    private AncestryDto ancestry;

    @JsonProperty("clazz")
    @JsonPropertyDescription("The class of this character")
    @Nullable
    private ClazzDto clazz;

    @JsonProperty("armor")
    @JsonPropertyDescription("The armor of this character")
    @Nullable
    private ArmorDto armor = null;

    @JsonProperty("shields")
    @JsonPropertyDescription("The shields of this character")
    @Nonnull
    private List<ShieldDto> shields = new ArrayList<>();

    @JsonProperty("weapons")
    @JsonPropertyDescription("The weapons of this character")
    @Nonnull
    private List<Weapon> weapons = new ArrayList<>();

    @JsonProperty("abilityScores")
    @JsonPropertyDescription("The ability scores of this character")
    @Nonnull
    Map<AbilityScoreEnum, AbilityScoreDto> abilityScores = Map.ofEntries(
            Map.entry(AbilityScoreEnum.STR, new AbilityScoreDto()),
            Map.entry(AbilityScoreEnum.DEX, new AbilityScoreDto()),
            Map.entry(AbilityScoreEnum.CON, new AbilityScoreDto()),
            Map.entry(AbilityScoreEnum.INT, new AbilityScoreDto()),
            Map.entry(AbilityScoreEnum.WIS, new AbilityScoreDto()),
            Map.entry(AbilityScoreEnum.CHA, new AbilityScoreDto())
    );

    @JsonProperty("skills")
    @JsonPropertyDescription("The skills of this character")
    @Nonnull
    Map<SkillEnum, ProficiencyBasedRollDto> skills = Map.ofEntries(
            Map.entry(SkillEnum.ACROBATICS, new ProficiencyBasedRollDto(AbilityScoreEnum.DEX)),
            Map.entry(SkillEnum.ARCANA, new ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            Map.entry(SkillEnum.ATHLETICS, new ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            Map.entry(SkillEnum.CRAFTING, new ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            Map.entry(SkillEnum.DECEPTION, new ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            Map.entry(SkillEnum.DIPLOMACY, new ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            Map.entry(SkillEnum.INTIMIDATION, new ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            Map.entry(SkillEnum.MEDICINE, new ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            Map.entry(SkillEnum.NATURE, new ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            Map.entry(SkillEnum.OCCULTISM, new ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            Map.entry(SkillEnum.PERFORMANCE, new ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            Map.entry(SkillEnum.RELIGION, new ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            Map.entry(SkillEnum.SOCIETY, new ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            Map.entry(SkillEnum.STEALTH, new ProficiencyBasedRollDto(AbilityScoreEnum.DEX)),
            Map.entry(SkillEnum.SURVIVAL, new ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            Map.entry(SkillEnum.THIEVERY, new ProficiencyBasedRollDto(AbilityScoreEnum.DEX))
    );

    @JsonProperty("savingThrows")
    @JsonPropertyDescription("The saving throws of this character")
    @Nonnull
    Map<SavingThrowEnum, ProficiencyBasedRollDto> savingThrows = Map.ofEntries(
            Map.entry(SavingThrowEnum.FORTITUDE, new ProficiencyBasedRollDto(AbilityScoreEnum.CON)),
            Map.entry(SavingThrowEnum.REFLEX, new ProficiencyBasedRollDto(AbilityScoreEnum.DEX)),
            Map.entry(SavingThrowEnum.WILL, new ProficiencyBasedRollDto(AbilityScoreEnum.WIS))
    );

    @JsonProperty("perception")
    @JsonPropertyDescription("The perception of this character")
    private ProficiencyBasedRollDto perception = new ProficiencyBasedRollDto(AbilityScoreEnum.WIS);

    @JsonProperty("classDC")
    @JsonPropertyDescription("The class DC of this character")
    private ProficiencyBasedRollDto classDC = new ProficiencyBasedRollDto();

    @JsonProperty("spellAttackRoll")
    @JsonPropertyDescription("The spell attack roll of this character")
    private ProficiencyBasedRollDto spellAttackRoll = new ProficiencyBasedRollDto();

    @JsonProperty("spellDC")
    @JsonPropertyDescription("The spell DC of this character")
    private ProficiencyBasedRollDto spellDC = new ProficiencyBasedRollDto();

    @JsonProperty("equipment")
    @JsonPropertyDescription("The equipment of this character")
    List<Equipment> equipment = new ArrayList<>();

    @JsonProperty("maxHitPoints")
    @JsonPropertyDescription("The maximum hit points of this character")
    private Integer maxHitPoints;

    @JsonProperty("currentHitPoints")
    @JsonPropertyDescription("The current hit points of this character")
    private Integer currentHitPoints;

    @JsonProperty("temporaryHitPoints")
    @JsonPropertyDescription("The temporary hit points of this character")
    private Integer temporaryHitPoints;

    @JsonProperty("speed")
    @JsonPropertyDescription("The speed of this character")
    private Integer speed;

    @JsonProperty("resistances")
    @JsonPropertyDescription("The resistances of this character")
    private List<DamageTypeEnum> resistances = new ArrayList<>();

    @JsonProperty("immunities")
    @JsonPropertyDescription("The immunities of this character")
    private List<DamageTypeEnum> immunities = new ArrayList<>();

    @JsonProperty("conditions")
    @JsonPropertyDescription("The conditions of this character")
    private List<DamageTypeEnum> conditions = new ArrayList<>();

    @JsonProperty("attackProficiencies")
    @JsonPropertyDescription("The attack proficiencies of this character")
    Map<AttackProficiencyEnum, ProficiencyDto> attackProficiencies = Map.ofEntries(
        Map.entry(MeleeWeaponTypeEnum.SWORD, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.SPEAR, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.HAMMER, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.POLEARM, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.KNIFE, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.CLUB, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.FLAIL, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.BRAWLING, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.AXE, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.PICK, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.SHIELD, new ProficiencyDto()),
        Map.entry(MeleeWeaponTypeEnum.WHIP, new ProficiencyDto()),
        Map.entry(RangedWeaponTypeEnum.DART, new ProficiencyDto()),
        Map.entry(RangedWeaponTypeEnum.BOW, new ProficiencyDto()),
        Map.entry(OtherAttackWeaponTypeEnum.UNARMED_ATTACKS, new ProficiencyDto()),
        Map.entry(OtherAttackWeaponTypeEnum.SIMPLE_WEAPONS, new ProficiencyDto()),
        Map.entry(OtherAttackWeaponTypeEnum.MARTIAL_WEAPONS, new ProficiencyDto()),
        Map.entry(OtherAttackWeaponTypeEnum.ALCHEMICAL_BOMBS, new ProficiencyDto())
    );

    @JsonProperty("defenseProficiencies")
    @JsonPropertyDescription("The defense proficiencies of this character")
    Map<DefenseProficiencyEnum, ProficiencyDto> defenseProficiencies = Map.ofEntries(
            Map.entry(DefenseProficiencyEnum.UNARMORED_DEFENSE, new ProficiencyDto()),
            Map.entry(DefenseProficiencyEnum.LIGHT_ARMOR, new ProficiencyDto()),
            Map.entry(DefenseProficiencyEnum.MEDIUM_ARMOR, new ProficiencyDto()),
            Map.entry(DefenseProficiencyEnum.HEAVY_ARMOR, new ProficiencyDto())
    );

    @JsonProperty("inventory")
    @JsonPropertyDescription("The inventory of this character")
    Map<InventoryItemEnum, String> inventory = new HashMap<>(); // items annotated by the granter

    public void clearItemsGrantedBy(String identifier) {
        List<InventoryItemEnum> toRemove = new ArrayList<>();
        for (Map.Entry<InventoryItemEnum, String> entry : this.inventory.entrySet()) {
            InventoryItemEnum key = entry.getKey();
            String grantedBy = entry.getValue();
            if (grantedBy.equals(identifier)) {
                toRemove.add(key);
            }
        }
        for (InventoryItemEnum itemEnum : toRemove) {
            this.inventory.remove(itemEnum);
        }
    }

    public void clearSkillsModifiersGrantedBy(String identifier) {
        for (Map.Entry<SkillEnum, ProficiencyBasedRollDto> entry : this.skills.entrySet()) {
            ProficiencyBasedRollDto skill = entry.getValue();
            skill.getProficiency().modifiers.removeIf(proficiencyModifierDto -> proficiencyModifierDto.granter.equals(identifier));
        }
    };
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/dto/CharacterDto.kt
package ai.ragu.charactermanager.dto

import ai.ragu.charactermanager.enumeration.*
import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyDescription
import com.fasterxml.jackson.annotation.ObjectIdGenerators.PropertyGenerator
import lombok.*
import java.util.*

@Data
@AllArgsConstructor
@NoArgsConstructor
@JsonIdentityInfo(generator = PropertyGenerator::class, property = "id")
@ToString(onlyExplicitlyIncluded = true)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
class CharacterDto {
    @JsonProperty("id")
    @JsonPropertyDescription("A unique identifier for this character")
    @ToString.Include
    @EqualsAndHashCode.Include
    @Nonnull
    private val id: UUID? = null
    @JsonProperty("name")
    @JsonPropertyDescription("The name of this character")
    @Nonnull
    private val name: String? = null
    @JsonProperty("level")
    @JsonPropertyDescription("The level of this character")
    @Nonnull
    private val level = 1
    @JsonProperty("ancestry")
    @JsonPropertyDescription("The ancestry of this character")
    private val ancestry: AncestryDto? = null
    @JsonProperty("clazz")
    @JsonPropertyDescription("The class of this character")
    private val clazz: ClazzDto? = null
    @JsonProperty("armor")
    @JsonPropertyDescription("The armor of this character")
    private val armor: ArmorDto? = null
    @JsonProperty("shields")
    @JsonPropertyDescription("The shields of this character")
    @Nonnull
    private val shields: List<ShieldDto> = ArrayList()
    @JsonProperty("weapons")
    @JsonPropertyDescription("The weapons of this character")
    @Nonnull
    private val weapons: List<Weapon> = ArrayList()
    @JsonProperty("abilityScores")
    @JsonPropertyDescription("The ability scores of this character")
    @Nonnull
    var abilityScores = java.util.Map.ofEntries(
            java.util.Map.entry(AbilityScoreEnum.STR, AbilityScoreDto()),
            java.util.Map.entry(AbilityScoreEnum.DEX, AbilityScoreDto()),
            java.util.Map.entry(AbilityScoreEnum.CON, AbilityScoreDto()),
            java.util.Map.entry(AbilityScoreEnum.INT, AbilityScoreDto()),
            java.util.Map.entry(AbilityScoreEnum.WIS, AbilityScoreDto()),
            java.util.Map.entry(AbilityScoreEnum.CHA, AbilityScoreDto())
    )
    @JsonProperty("skills")
    @JsonPropertyDescription("The skills of this character")
    @Nonnull
    var skills = java.util.Map.ofEntries(
            java.util.Map.entry(SkillEnum.ACROBATICS, ProficiencyBasedRollDto(AbilityScoreEnum.DEX)),
            java.util.Map.entry(SkillEnum.ARCANA, ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            java.util.Map.entry(SkillEnum.ATHLETICS, ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            java.util.Map.entry(SkillEnum.CRAFTING, ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            java.util.Map.entry(SkillEnum.DECEPTION, ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            java.util.Map.entry(SkillEnum.DIPLOMACY, ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            java.util.Map.entry(SkillEnum.INTIMIDATION, ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            java.util.Map.entry(SkillEnum.MEDICINE, ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            java.util.Map.entry(SkillEnum.NATURE, ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            java.util.Map.entry(SkillEnum.OCCULTISM, ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            java.util.Map.entry(SkillEnum.PERFORMANCE, ProficiencyBasedRollDto(AbilityScoreEnum.CHA)),
            java.util.Map.entry(SkillEnum.RELIGION, ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            java.util.Map.entry(SkillEnum.SOCIETY, ProficiencyBasedRollDto(AbilityScoreEnum.INT)),
            java.util.Map.entry(SkillEnum.STEALTH, ProficiencyBasedRollDto(AbilityScoreEnum.DEX)),
            java.util.Map.entry(SkillEnum.SURVIVAL, ProficiencyBasedRollDto(AbilityScoreEnum.WIS)),
            java.util.Map.entry(SkillEnum.THIEVERY, ProficiencyBasedRollDto(AbilityScoreEnum.DEX))
    )
    @JsonProperty("savingThrows")
    @JsonPropertyDescription("The saving throws of this character")
    @Nonnull
    var savingThrows = java.util.Map.ofEntries(
            java.util.Map.entry(SavingThrowEnum.FORTITUDE, ProficiencyBasedRollDto(AbilityScoreEnum.CON)),
            java.util.Map.entry(SavingThrowEnum.REFLEX, ProficiencyBasedRollDto(AbilityScoreEnum.DEX)),
            java.util.Map.entry(SavingThrowEnum.WILL, ProficiencyBasedRollDto(AbilityScoreEnum.WIS))
    )
    @JsonProperty("perception")
    @JsonPropertyDescription("The perception of this character")
    private val perception = ProficiencyBasedRollDto(AbilityScoreEnum.WIS)
    @JsonProperty("classDC")
    @JsonPropertyDescription("The class DC of this character")
    private val classDC = ProficiencyBasedRollDto()
    @JsonProperty("spellAttackRoll")
    @JsonPropertyDescription("The spell attack roll of this character")
    private val spellAttackRoll = ProficiencyBasedRollDto()
    @JsonProperty("spellDC")
    @JsonPropertyDescription("The spell DC of this character")
    private val spellDC = ProficiencyBasedRollDto()
    @JsonProperty("equipment")
    @JsonPropertyDescription("The equipment of this character")
    var equipment: List<Equipment> = ArrayList()
    @JsonProperty("maxHitPoints")
    @JsonPropertyDescription("The maximum hit points of this character")
    private val maxHitPoints: Int? = null
    @JsonProperty("currentHitPoints")
    @JsonPropertyDescription("The current hit points of this character")
    private val currentHitPoints: Int? = null
    @JsonProperty("temporaryHitPoints")
    @JsonPropertyDescription("The temporary hit points of this character")
    private val temporaryHitPoints: Int? = null
    @JsonProperty("speed")
    @JsonPropertyDescription("The speed of this character")
    private val speed: Int? = null
    @JsonProperty("resistances")
    @JsonPropertyDescription("The resistances of this character")
    private val resistances: List<DamageTypeEnum> = ArrayList()
    @JsonProperty("immunities")
    @JsonPropertyDescription("The immunities of this character")
    private val immunities: List<DamageTypeEnum> = ArrayList()
    @JsonProperty("conditions")
    @JsonPropertyDescription("The conditions of this character")
    private val conditions: List<DamageTypeEnum> = ArrayList()
    @JsonProperty("attackProficiencies")
    @JsonPropertyDescription("The attack proficiencies of this character")
    var attackProficiencies: Map<AttackProficiencyEnum, ProficiencyDto> = java.util.Map.ofEntries(
            java.util.Map.entry(MeleeWeaponTypeEnum.SWORD, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.SPEAR, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.HAMMER, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.POLEARM, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.KNIFE, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.CLUB, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.FLAIL, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.BRAWLING, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.AXE, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.PICK, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.SHIELD, ProficiencyDto()),
            java.util.Map.entry(MeleeWeaponTypeEnum.WHIP, ProficiencyDto()),
            java.util.Map.entry(RangedWeaponTypeEnum.DART, ProficiencyDto()),
            java.util.Map.entry(RangedWeaponTypeEnum.BOW, ProficiencyDto()),
            java.util.Map.entry(OtherAttackWeaponTypeEnum.UNARMED_ATTACKS, ProficiencyDto()),
            java.util.Map.entry(OtherAttackWeaponTypeEnum.SIMPLE_WEAPONS, ProficiencyDto()),
            java.util.Map.entry(OtherAttackWeaponTypeEnum.MARTIAL_WEAPONS, ProficiencyDto()),
            java.util.Map.entry(OtherAttackWeaponTypeEnum.ALCHEMICAL_BOMBS, ProficiencyDto())
    )
    @JsonProperty("defenseProficiencies")
    @JsonPropertyDescription("The defense proficiencies of this character")
    var defenseProficiencies = java.util.Map.ofEntries(
            java.util.Map.entry(DefenseProficiencyEnum.UNARMORED_DEFENSE, ProficiencyDto()),
            java.util.Map.entry(DefenseProficiencyEnum.LIGHT_ARMOR, ProficiencyDto()),
            java.util.Map.entry(DefenseProficiencyEnum.MEDIUM_ARMOR, ProficiencyDto()),
            java.util.Map.entry(DefenseProficiencyEnum.HEAVY_ARMOR, ProficiencyDto())
    )
    @JsonProperty("inventory")
    @JsonPropertyDescription("The inventory of this character")
    var inventory: Map<InventoryItemEnum, String> = HashMap() // items annotated by the granter

    fun clearItemsGrantedBy(identifier: String) {
        val toRemove: MutableList<InventoryItemEnum> = ArrayList()
        for ((key, grantedBy) in inventory) {
            if (grantedBy == identifier) {
                toRemove.add(key)
            }
        }
        for (itemEnum in toRemove) {
            inventory.remove(itemEnum)
        }
    }

    fun clearSkillsModifiersGrantedBy(identifier: String) {
        for ((_, skill) in skills) {
            skill.proficiency.modifiers.removeIf { proficiencyModifierDto: ProficiencyModifierDto -> proficiencyModifierDto.granter == identifier }
        }
    }
}